### PR TITLE
Fixed bugs, typos

### PR DESCRIPTION
**Pull request by WinterSilence**
fixed bug with undefined _protocol, moved from protocol() to __construct() - it will be faster

`return UTF8::strlen($this->_body);` correctly than strlen()

`if (function_exists('http_build_cookie'))` [function_exists faster than extension_loaded](http://www.merkushin.com/function_exists-vs-extension_loaded/) sick! but true 0_o

### CODE REVIEW

File: classes/Kohana/Response.php [hunk 1]
```diff
@@ -1,29 +1,27 @@
...
  * @category   Base
  * @author     Kohana Team
- * @copyright  (c) 2008-2012 Kohana Team
- * @license    http://kohanaphp.com/license
- * @since      3.1.0
+ * @copyright  (c) 2008-2014 Kohana Team
+ * @license    http://kohanaframework.org/license
  */
-class Kohana_Response implements HTTP_Response {
+abstract class Kohana_Response implements HTTP_Response {
```
Comment (shadowhand) [inline classes/Kohana/Response.php:12]: what... this makes no sense.

File: classes/Kohana/Response.php [hunk 2]
```diff
@@ -1,29 +1,27 @@
...
 
 	/**
 	 * Factory method to create a new [Response]. Pass properties
 	 * in using an associative array.
 	 *
-	 *      // Create a new response
-	 *      $response = Response::factory();
+	 *     // Create a new response
+	 *     $response = Response::factory();
```
Comment (shadowhand) [inline classes/Kohana/Response.php:19]: why are you reformatting whitespace?

File: classes/Kohana/Response.php [hunk 3]
```diff
@@ -97,44 +97,48 @@ public static function factory(array $config = array())
...
-				{
-					$this->$key = $value;
-				}
+				$this->$key = $value;
 			}
 		}
+
+		if (!$this->_protocol)
+		{
+			$this->_protocol = HTTP::$protocol;
```
Comment (shadowhand) [inline classes/Kohana/Response.php:140]: and where does `$protocol` come from? it's not defined...

File: classes/Kohana/Response.php [hunk 4]
```diff
@@ -176,30 +179,25 @@ public function protocol($protocol = NULL)
 			return $this;
 		}
 
-		if ($this->_protocol === NULL)
```
Comment (shadowhand) [inline classes/Kohana/Response.php:179]: looks like that was a mistake...

File: classes/Kohana/Response.php [hunk 5]
```diff
@@ -176,30 +179,25 @@ public function protocol($protocol = NULL)
...
+	 *     // Get the current status
+	 *     $status = $response->status();
 	 *
-	 * @param   integer  $status Status to set to this response
+	 * @param   integer  $status  Status to set to this response
 	 * @return  mixed
+	 * @throws  Kohana_Exception
 	 */
-	public function status($status = NULL)
+	public function status($status = 0)
```
Comment (shadowhand) [inline classes/Kohana/Response.php:198]: why?

File: classes/Kohana/Response.php [hunk 6]
```diff
@@ -210,7 +208,10 @@ public function status($status = NULL)
 		}
 		else
 		{
-			throw new Kohana_Exception(__METHOD__.' unknown status value : :value', array(':value' => $status));
+			throw new Kohana_Exception(
```
Comment (shadowhand) [inline classes/Kohana/Response.php:211]: why?

File: classes/Kohana/Response.php [hunk 7]
```diff
@@ -248,24 +249,21 @@ public function headers($key = NULL, $value = NULL)
 		}
 		elseif ($value === NULL)
 		{
-			return Arr::get($this->_header, $key);
-		}
-		else
-		{
-			$this->_header[$key] = $value;
-			return $this;
+			return $this->_header->offsetGet($value);
```
Comment (shadowhand) [inline classes/Kohana/Response.php:252]: why

File: classes/Kohana/Response.php [hunk 8]
```diff
@@ -248,24 +249,21 @@ public function headers($key = NULL, $value = NULL)
...
-	 * Returns the length of the body for use with
-	 * content header
+	 * Returns the length of the body for use with content header
 	 *
 	 * @return  integer
 	 */
 	public function content_length()
 	{
-		return strlen($this->body());
+		return UTF8::strlen($this->_body);
```
Comment (shadowhand) [inline classes/Kohana/Response.php:266]: ~~finally! a good change.~~

edit: nope, see next comment.

File: classes/Kohana/Response.php [hunk 9]
```diff
@@ -379,55 +376,37 @@ public function send_headers($replace = FALSE, $callback = NULL)
...
+			$size = $this->content_length();
 
 			// Create a temporary file to hold the current response
 			$file = tmpfile();
 
 			// Write the current response into the file
-			fwrite($file, $file_data);
-
-			// File data is no longer needed
-			unset($file_data);
```
Comment (shadowhand) [inline classes/Kohana/Response.php:430]: what... why?

File: classes/Kohana/Response.php [hunk 10]
```diff
@@ -443,34 +422,29 @@ public function send_file($filename, $download = NULL, array $options = NULL)
 			// Get the file size
 			$size = filesize($filename);
 
-			if ( ! isset($mime))
```
Comment (shadowhand) [inline classes/Kohana/Response.php:446]: wtf...

File: classes/Kohana/Response.php [hunk 11]
```diff
@@ -443,34 +422,29 @@ public function send_file($filename, $download = NULL, array $options = NULL)
...
-			));
+			throw new Kohana_Exception(
+				':method: Could not read file to send: :file', 
+				array(':method' => __METHOD__, ':file' => $download)
+			);
 		}
 
 		// Inline or download?
-		$disposition = empty($options['inline']) ? 'attachment' : 'inline';
+		$disposition = isset($options['inline']) ? 'inline' : 'attachment';
```
Comment (shadowhand) [inline classes/Kohana/Response.php:438]: what if i pass `inline = FALSE` ?

File: classes/Kohana/Response.php [hunk 12]
```diff
@@ -443,34 +422,29 @@ public function send_file($filename, $download = NULL, array $options = NULL)
...
 
 		// Calculate byte range to download.
 		list($start, $end) = $this->_calculate_byte_range($size);
 
-		if ( ! empty($options['resumable']))
+		if (isset($options['resumable']))
```
Comment (shadowhand) [inline classes/Kohana/Response.php:443]: same thing, what about `resumable = false` ?

File: classes/Kohana/Response.php [hunk 13]
```diff
@@ -479,23 +453,34 @@ public function send_file($filename, $download = NULL, array $options = NULL)
...
 			// Naturally, IE does not act like a real browser...
-			if (Request::$initial->secure())
+			if (Request::initial()->secure())
 			{
-				// http://support.microsoft.com/kb/316431
+				// @link http://support.microsoft.com/kb/316431
 				$this->_header['pragma'] = $this->_header['cache-control'] = 'public';
 			}
 
-			if (version_compare(Request::user_agent('version'), '8.0', '>='))
```
Comment (shadowhand) [inline classes/Kohana/Response.php:496]: wtf..

File: classes/Kohana/Response.php [hunk 14]
```diff
@@ -503,10 +488,12 @@ public function send_file($filename, $download = NULL, array $options = NULL)
 		// Send all headers now
 		$this->send_headers();
 
-		while (ob_get_level())
+		// Flush all output buffers
+		$levels = ob_get_level();
```
Comment (shadowhand) [inline classes/Kohana/Response.php:492]: uh, no...

File: classes/Kohana/Response.php [hunk 15]
```diff
@@ -518,16 +505,19 @@ public function send_file($filename, $download = NULL, array $options = NULL)
...
 
-		// Send data in 16kb blocks
-		$block = 1024 * 16;
+		// Send data in 16kb (1024*16) blocks
+		$block = 16384;
 
 		fseek($file, $start);
 
-		while ( ! feof($file) AND ($pos = ftell($file)) <= $end)
+		while ( ! connection_aborted() AND ! feof($file))
```
Comment (shadowhand) [inline classes/Kohana/Response.php:513]: again, no.

File: classes/Kohana/Response.php [hunk 16]
```diff
@@ -544,7 +534,7 @@ public function send_file($filename, $download = NULL, array $options = NULL)
 		// Close the file
 		fclose($file);
 
-		if ( ! empty($options['delete']))
+		if (isset($options['delete']) AND $options['delete'])
```
Comment (shadowhand) [inline classes/Kohana/Response.php:537]: wow, pointless changes abound.

File: classes/Kohana/Response.php [hunk 17]
```diff
@@ -556,13 +546,10 @@ public function send_file($filename, $download = NULL, array $options = NULL)
 				// Create a text version of the exception
 				$error = Kohana_Exception::text($e);
 
-				if (is_object(Kohana::$log))
+				if (Kohana::$log instanceof Log)
```
Comment (shadowhand) [inline classes/Kohana/Response.php:549]: no.

**Comment by shadowhand** [comment]
i stopped reviewing half way through. if you can't separate your opinions from your bug fixes, please stop submitting pull requests. it is a huge waste of everyone's time when you do this.

**Comment by WinterSilence** [comment]
@shadowhand Nobody forces to spend your time, you can ignore my requests. Some of the issues you mentioned, I described in the comments to request.
I do not understand why asking questions and then close the query, you probably are not interested in the answers.

**Comment by shadowhand** [comment]
@WinterSilence i should ignore pull requests? okay... i'll quote you on that next time someone complains that no one pays attention to them.

File: classes/Kohana/Response.php (hunk 1 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:12]: Then half of the available classes.
Eg: `abstract class Kohana_Session {`
https://github.com/kohana/core/blob/3.4/develop/classes/Kohana/Session.php#L11

File: classes/Kohana/Response.php (hunk 2 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:19]: remove extra spaces, should be 4 spaces for code + 1 border

Comment (shadowhand) [inline classes/Kohana/Response.php:19]: so what? it has **nothing** to do with the pull request goals of "fixed bugs"

File: classes/Kohana/Response.php (hunk 3 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:140]: At current version default protocol is set only when calling `protocol()`, this wrong

File: classes/Kohana/Response.php (hunk 2 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:19]: I'm trying to tidy up the code and comments to him. Request title: "Fixed bugs, typos"

File: classes/Kohana/Response.php (hunk 4 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:179]: See latest version https://github.com/WinterSilence/core/blob/9e6d8368d4cfde907706ccf191ce7f45a01fb99b/classes/Kohana/Response.php#L174

File: classes/Kohana/Response.php (hunk 2 above)
Comment (shadowhand) [inline classes/Kohana/Response.php:19]: each bug fix should be a separate pull request, as a **clear**, **single** unit of change. you can put all of your opinion changes (like all the bogus whitespace changes) into a separate pull request.

File: classes/Kohana/Response.php (hunk 5 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:198]: `@param   integer  $status` NULL not integer value

Comment (shadowhand) [inline classes/Kohana/Response.php:198]: yes, but `function (array $foo = NULL)` is totally acceptable. NULL literally means "not defined" and is an acceptable value for any **type** of optional parameter.

this fixes nothing and adds absolutely no value.

File: classes/Kohana/Response.php (hunk 6 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:211]: Cut long string, corrected passing arguments

File: classes/Kohana/Response.php (hunk 7 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:252]: Why native method better? are you serious?

Comment (shadowhand) [inline classes/Kohana/Response.php:252]: no, you modified the logic, for no apparent reason... why did you modify the if/else statements? this is not fixing anything, it is an opinion change.

File: classes/Kohana/Response.php (hunk 9 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:430]: Extra variables are not needed

File: classes/Kohana/Response.php (hunk 10 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:446]: Merge duplicate code and posted below.

File: classes/Kohana/Response.php (hunk 11 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:438]: what if you pass nothing? - notice. I remove an additional check because the value is not used, its only flag

File: classes/Kohana/Response.php (hunk 12 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:443]: By analogy to `$options['inline']`

File: classes/Kohana/Response.php (hunk 13 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:496]: I already wrote that this version is faster.

File: classes/Kohana/Response.php (hunk 14 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:492]: > If you enable zlib.output_compression then level count will be increased by 1 and then this code will just freeze your script.
> http://ru2.php.net/manual/ru/function.ob-end-flush.php#103387

File: classes/Kohana/Response.php (hunk 15 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:513]: Explain your answer

File: classes/Kohana/Response.php (hunk 17 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:549]: Explain your answer

File: classes/Kohana/Response.php (hunk 14 above)
Comment (Ikke) [inline classes/Kohana/Response.php:492]: That's why you should create individual commits that mention these kinds of
details, not one mass change commit that is difficult to accept while there
are still issues and conveys very little information about why certain
changes are being made, especially when looking back at the commits.

Comment (Ikke) [inline classes/Kohana/Response.php:492]: My bad, I should have watched better, there are separate commits.

File: classes/Kohana/Response.php (hunk 5 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:198]: For "strong" typing is not the correct value. Just look listing functions and methods of php

File: classes/Kohana/Response.php (hunk 11 above)
Comment (shadowhand) [inline classes/Kohana/Response.php:438]: by "pass nothing" i assume you mean `$options` being null, or `$options['inline']` not being defined. but in reality, [empty](http://php.net/empty) works like `isset` and doesn't produce notices or warnings for undefined array offsets, as per the manual:

>  That means empty() is essentially the concise equivalent to !isset($var) || $var == false.

File: classes/Kohana/Response.php (hunk 13 above)
Comment (shadowhand) [inline classes/Kohana/Response.php:496]: that's not a bug fix or a type, that's a performance change. hence, should be a separate pull request.

File: classes/Kohana/Response.php (hunk 14 above)
Comment (shadowhand) [inline classes/Kohana/Response.php:492]: No, his pull request message mentions none of this. You're absolutely right that each of these should be a separate pull request!

File: classes/Kohana/Response.php (hunk 11 above)
Comment (WinterSilence) [inline classes/Kohana/Response.php:438]: This is only true in this case as after checking the option not used. Сейчас добавлю для этих опций дополнительную проверку

Comment (shadowhand) [inline classes/Kohana/Response.php:438]: You're wrong:

``` php
<?php
error_reporting(E_ALL);
// this produces no notice, warning, or error
var_dump(empty($foo["bar"]));
```

**Comment by WinterSilence** [comment]
I am engaged my fork, if I find any errors or optimize code, then I send it here too. I don't have time for a more detailed request, if something looks interesting, you can add its own separate request.

**Comment by shadowhand** [comment]
> I don't have time for a more detailed request, if something looks interesting, you can add its own separate request.

You are acting as if your time is more valuable and important than anyone else's. It would have taken you matter of minutes to separate these commits into distinct pull requests, but you instead you argued with my review.

I appreciate that you are trying to help, but if you would just focus your effort on using the correct process, rather than arguing, we would be able to use your contributions.

**Comment by WinterSilence** [comment]
judging by the fact that a lot of criticism, and no help, then is not true. if no one wants to develop a project, makes no sense to send new requests, rather spend this time on my fork.

**Comment by shadowhand** [comment]
> judging by the fact that a lot of criticism, and no help

I made it very clear what the issues with this pull request were. No one will collaborate with you on a project if you don't follow the process or make any attempt to address concerns.

File: classes/Kohana/Response.php (hunk 8 above)
Comment (cs278) [inline classes/Kohana/Response.php:266]: Except it's wrong, the content length header should express the [number of bytes not the number of characters](http://mark.koli.ch/remember-kids-an-http-content-length-is-the-number-of-bytes-not-the-number-of-characters).

Comment (shadowhand) [inline classes/Kohana/Response.php:266]: good catch chris!

Comment (WinterSilence) [inline classes/Kohana/Response.php:266]: @cs278 :+1:
